controllers/service_cache.go

```diff
@@ -0,0 +1,111 @@
+package controllers
+
+import (
+	"context"
+	stderrors "errors"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"slime.io/slime/framework/util"
+	"sync"
+)
+
+func newLabelSvcCache(clientSet *kubernetes.Clientset) (map[string][]string, *sync.RWMutex, error) {
+	log := log.WithField("function", "newLabelSvcCache")
+	labelSvcCache := make(map[string][]string)
+	var cacheLock sync.RWMutex
+
+	// init labelSvcCache
+	services, err := clientSet.CoreV1().Services("").List(metav1.ListOptions{})
+	if err != nil {
+		return nil, nil, stderrors.New("failed to get service list")
+	}
+
+	for _, service := range services.Items {
+		svc := service.GetNamespace() + "/" + service.GetName()
+		var labels []string
+		for k, v := range service.GetLabels() {
+			label := k+"="+v
+			labels = append(labels, label)
+			labelSvcCache[label]=append(labelSvcCache[label], svc)
+		}
+	}
+
+	// init service watcher
+	servicesClient := clientSet.CoreV1().Services("")
+	lw := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			return servicesClient.List(options)
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			return servicesClient.Watch(options)
+		},
+	}
+	watcher := util.ListWatcher(context.Background(), lw)
+
+	go func() {
+		log.Infof("Service cacher is running")
+		for {
+			e, ok := <-watcher.ResultChan()
+			if !ok {
+				log.Warningf("a result chan of service watcher is closed, break process loop")
+				return
+			}
+
+			service, ok := e.Object.(*v1.Service)
+			if !ok {
+				log.Errorf("invalid type of object in service watcher event")
+				continue
+			}
+
+			eventSvc := service.GetNamespace() + "/" + service.GetName()
+			// delete event
+			if e.Type == watch.Deleted {
+				cacheLock.Lock()
+				for k, v := range service.GetLabels() {
```

Comment:
we'd better remove the svc from all label-items to keep a better final-consistency.